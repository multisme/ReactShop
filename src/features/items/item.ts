import { createSlice } from "@reduxjs/toolkit";
import { AppThunk } from "../../app/store";

export interface itemData {
  id: number | undefined;
  quantity: number;
  name: string;
}

export const initialState = {
  id: undefined,
  quantity: 0,
  name: "unknown",
} as itemData;

const ItemSlice = createSlice({
  name: "item",
  initialState,
  reducers: {},
});
