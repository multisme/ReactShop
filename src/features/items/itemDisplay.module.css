.productList{
        list-style: none;
        display: flex;
}

.product{
        width: 150px;
        height: 150px;
}

.centeredContent{
        display: flex;
        align-items: center;
        justify-content: center;
}
