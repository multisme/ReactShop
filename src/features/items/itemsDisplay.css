.itemsList{
        list-style: none;
        display: flex;
        flex-wrap: wrap;
        margin: 0.25rem;
        padding: 0px;
}
