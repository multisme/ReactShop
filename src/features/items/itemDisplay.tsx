import React from "react";
import {useDispatch} from "react-redux"

import { itemData } from "features/items/item";
import { updateSelectedItem } from "features/showcase/showcaseSlice"

import styles from 'features/items/itemDisplay.module.css';


interface Props{
        items: itemData[]
}


export const ItemListDisplay = ({ items }: Props) => {
        const renderedlist = items.map((item: itemData) => <ItemDisplay {...item}/>);
        return <ul className={styles.productList} id="list">{renderedlist}</ul>;
};

export const ItemDisplay = (item: itemData) => {
//change the event type here with the correct one

const dispatch = useDispatch();

const selectItem = (e: any) => {
        e.preventDefault();
       dispatch( updateSelectedItem(item) );
}

  return (
    <li
        onClick={selectItem}
        className={styles.product} key={item.id}
    >
      <div className="name">{item.name}</div>
      <div className={"quantity"}>{item.quantity}</div>
    </li>
  );
};
