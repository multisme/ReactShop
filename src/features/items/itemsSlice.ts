import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppThunk } from "app/store";
import { itemData } from "features/item/itemSlice";

export interface itemListState {
  loading: boolean;
  hasError: boolean;
  items: itemData[];
}

const initialState = {
  loading: false,
  hasError: false,
  items: [],
} as itemListState;

const ItemsSlice = createSlice({
  name: "items",
  initialState,
  reducers: {
    getItems: (state) => {
      state.loading = true;
    },
    getItemsSuccess: (state, { payload }: PayloadAction<itemData[]>) => {
      state.loading = false;
      state.items = payload;
      state.hasError = false;
    },
    getItemsFailure: (state) => {
      state.loading = false;
      state.hasError = true;
    },
  },
});

// The Actions
export const {
  getItems,
  getItemsSuccess,
  getItemsFailure,
} = ItemsSlice.actions;

export const itemsSelector = (state: { items: itemListState }) =>
  state.items;

export const itemPageSelector = createSelector(
        (state) => state.items.items,
        (_: any, id: string | undefined) => id,
        (items, id) => items.filter((item: itemData) => item.id == id)
)

//Thunk action TO fetch the Items
export const fetchItems = (): AppThunk => async (dispatch) => {
  dispatch(getItems());

  try {
    const response = await fetch("http://localhost:3000/items");

    const data = await response.json();
    dispatch(getItemsSuccess(data));
  } catch (error) {
    dispatch(getItemsFailure());
  }
};

//The reducer
export default ItemsSlice.reducer;
