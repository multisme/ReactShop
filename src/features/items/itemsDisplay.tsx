import React from "react";

import {useDispatch} from "react-redux"
import {Link} from "react-router-dom"

import { itemData } from "features/item/itemSlice";
import {
        updateSelectedItem,
        removeSelectedItem
} from "features/showcase/showcaseSlice";

import "features/items/itemsDisplay.css";

interface itemListProps{
items: itemData[]
}

interface itemProps{
item: itemData
}

const iconPath = process.env.PUBLIC_URL + '/assets/images/';

console.log("cin", iconPath);
export const ItemListDisplay = ({ items }: itemListProps) => {
        const renderedlist = items.map((item: itemData) =>
                        <div className="itemCard"  key={item.id} style={{backgroundImage: "url("  +item.url +")"}}>
                        <Link key={item.id}to={`/products/${item.id}`}>
                        <ItemDisplay item={item}/>
                        </Link>
                        </div>
                        );
        return (
                        <ul className="itemsList" id="list">
                        {renderedlist}
                        </ul>
               );
};

export const ItemDisplay = ({item}: itemProps) => {
        //change the event type here with the correct one

        const dispatch = useDispatch();

        const selectItem = (e: any) => {
                console.log("mouse enter");
                e.preventDefault();
                dispatch( updateSelectedItem(item) );
        }

        const unselectItem = (e: any) => {
                dispatch( removeSelectedItem() );
        }

        return (
                        <li
                        onMouseEnter={selectItem}
                        onMouseLeave={unselectItem}
                        className={"flex-centered"}
                        >
                        <div className="picture">
                        </div>
                        <div className={"detailsSmall flex-centered"}>
                        <div className="name">{item.name}</div>
                        <div className={"price"}>{item.price}â‚¬</div>
                        </div>
                        </li>
               );
};
