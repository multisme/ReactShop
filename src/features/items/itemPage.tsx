import React from "react";
import { useDispatch, useSelector } from "react-redux";
import {Link} from "react-router-dom";
import {
        useHistory,
        useParams
 } from "react-router-dom";

import { RootState } from "app/rootReducer";
import {itemData} from "features/item/itemSlice";
import {itemPageSelector} from "features/items/itemsSlice";
import {addToCart} from "features/cart/cartSlice";

import { useFormFields } from "utils/utils";

interface productPageParams{
        id: string
}

const ItemPage = ()=>{
        const {id}  = useParams<productPageParams>();
        const [fields, handleFieldChange] = useFormFields({
                quantity: 0
        });
        const dispatch = useDispatch();
        const [item] = useSelector((state) =>itemPageSelector(state, id));
        const history = useHistory();
        if (id == undefined || item == undefined){
                return (
                <h3>Error</h3>
                );
        }
       
        const handleSubmit = (e: any) => {
                e.preventDefault();
                if (fields.quantity.content > 0){
                        dispatch(addToCart({
                        id: parseInt(id),
                        quantity: fields.quantity.content,
                        price: item.price}));
                        history.push("/bill")
                } else {
                        alert("Please choose something before checkout");
                }
        }
        
        const range = [...Array(item.quantity + 1).keys()]
        const availableQuantities = range.map((i) => 
                <option value={i} key={i}>{i}</option>
        )
 
        return (
        <div className="itemPage flex-centered">
                <div className="panel left flex-centered">
                <div className="mainImage">
                <img src="https://teamtijger.nl/wp-content/uploads/2018/06/zwarte-trui-transparant-300x300.png)" />;
                </div>
                </div>
                <div className="panel right">
                        <div className="info">
                                <div className="name">
                                {item.name}
                                </div>
                                <div className="price">
                                {item.price}
                                </div>
                                <div className="details">
                                {item.details}
                                </div>
                        </div>
                <form className="itemForm" onSubmit={handleSubmit}>
                <fieldset className="quantity">
                <select name="quantity" id="quantity" onChange={handleFieldChange}>
                        {availableQuantities}
                </select>
                </fieldset>
                <fieldset className="button">
                <Link to="/">
                <button id="goBack" className="buy" >
                        GO BACK
                </button>
                </Link>
                <button type="submit" id="checkout" className="checkout">
                        CHECKOUT
                </button>
                </fieldset>
                </form>
                </div>
        </div>
        )
}
export default ItemPage;
