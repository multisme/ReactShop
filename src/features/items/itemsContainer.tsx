import React from "react";
import { useSelector } from "react-redux";

import { itemListSelector } from "features/items/itemsSlice";

import { ItemListDisplay } from "features/items/itemsDisplay";

import "features/items/itemsContainer.css";

export const ItemsContainer = () => {

  const { hasError, loading, items } = useSelector(itemListSelector);
 
  if (hasError) {
    return (
      <div style={{fontSize: "30px"}}>
        <h1>Something went wrong with the backend</h1>
      </div>
    );
  }

  let renderedContainer = loading ? (
    <h3>Loading...</h3>
  ) : (
    <div className="itemListWrapper">
        <ItemListDisplay items={items} />
    </div>
  );

  return <div className="ListContainer">{renderedContainer}</div>;
};

export default ItemsContainer;
