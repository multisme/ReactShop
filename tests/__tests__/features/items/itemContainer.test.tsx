import React from 'react';
import configureStore from 'redux-mock-store';
import * as redux from 'react-redux';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { BrowserRouter as Router} from 'react-router-dom';

import { ItemsContainer } from "features/items/itemsContainer";
import { ItemListDisplay } from "features/items/itemsDisplay";

import {
        getItems,
        getItemsSuccess,
        getItemsFailure
 } from "features/items/itemsSlice";

const mockStore = configureStore([]);

describe('itemsContainer', () => {
        let store: any;
        const dispatch = jest.fn();
                  beforeEach(() => {
                          store = mockStore({
                  })
                })

                it("correct component rendered on instanciation", () => {
                        const mock = jest.spyOn(redux, 'useSelector');
                        mock.mockImplementation(() => ({ 
                                loading: true,
                                hasError: false,
                                items: []
                        }))

                        const component = mount(
                                <Provider store={store}>
                                        <ItemsContainer />
                                </Provider>
                        );
                        const h3 = component.find('h3');
                        expect(h3.length).toEqual(1);
                        expect(component).toMatchSnapshot();
                        expect(h3.text()).toEqual('Loading...')
                })
                it("correct component rendered on hasError", () => {
                        const mock = jest.spyOn(redux, 'useSelector');
                        mock.mockImplementation(() => ({ 
                                loading: false,
                                hasError: true,
                                items: []
                        }))

                        const component = mount(
                                <redux.Provider store={store}>
                                        <ItemsContainer />
                                </redux.Provider>
                        );
                        const h1 = component.find('h1');
                        const h3 = component.find('h3');
                        expect(h1.length).toEqual(1);
                        expect(h3.length).toEqual(1);
                        expect(component).toMatchSnapshot();
                        expect(h1.text()).toEqual('Something went wrong with the database')
                        expect(h3.text()).toEqual('Is the developper payed well enough?')
                })
                it("correct component rendered", () => {
                        const mock = jest.spyOn(redux, 'useSelector');
                        mock.mockImplementation(() => ({ 
                                loading: false,
                                hasError: false,
                                items: [
                                {id: 3, name: "dede", quantity: 4, price: 5},
                                {id: 6, name: "dodo", quantity: 2, price: 8}
                                ]
                        }))

                        const component = mount(
                                <Router>
                                <redux.Provider store={store}>
                                        <ItemsContainer />
                                </redux.Provider>
                                </Router>
                        );
                        expect(component.find('.itemCard')).toHaveLength(2);
                })
})
