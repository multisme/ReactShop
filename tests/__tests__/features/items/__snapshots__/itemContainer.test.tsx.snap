// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`itemsContainer correct component rendered on hasError 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ItemsContainer>
    <div
      style={
        Object {
          "fontSize": "30px",
        }
      }
    >
      <h1>
        Something went wrong with the backend
      </h1>
    </div>
  </ItemsContainer>
</Provider>
`;

exports[`itemsContainer correct component rendered on instanciation 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ItemsContainer>
    <div
      className="ListContainer"
    >
      <h3>
        Loading...
      </h3>
    </div>
  </ItemsContainer>
</Provider>
`;
